src/draw/ElkLayout.js

```diff
@@ -0,0 +1,344 @@
+   */
+  static __elk = new ELK();
+
+  /**
+   * @param {DefaultData} pluginData - a graph to be arranged.
+   * @param {ElkParams} [elkParams] - Parameters for the layout algorithm. (WIP, use defaults)
+   */
+  constructor(pluginData, elkParams) {
+    super(pluginData);
+
+    // default parameters
+    this.elkParams = {
+      interactive: false,
+      separateConnectedComponents: true,
+      cycleBreaking: 'INTERACTIVE',
+      layering: 'INTERACTIVE',
+      crossingMinimization: 'LAYER_SWEEP',
+      nodePlacement: 'BRANDES_KOEPF',
+      interactiveReferencePoint: 'TOP_LEFT',
+      direction: 'UNDEFINED',
+
+      ...elkParams,
+    };
+  }
+
+  /**
+   * Automatically arrange components.
+   * @returns {Promise<void>} Promise with nothing on success otherwise an error.
+   * @public
+   */
+  async arrangeComponentsPosition() {
+    const { components } = this.pluginData;
+    const links = this.pluginData.getLinks();
+
+    const layout = await this.__generateAllElkLayouts(components, links);
+
+    this.__writeLayout(layout);
+  }
+
+  /**
+   * Generate automatic layout for a given graph.
+   * @param {Component[]} components - The components to be automatically arranged.
+   * @param {ComponentLink[]} links - Links between the given components.
+   * @returns {Promise<ElkNode[]>} A layout for the graph, see {@link __writeLayout}.
+   * @private
+   */
+  async __generateAllElkLayouts(components, links) {
+    // get tree nodes from components (each node represents a single component and its relations)
+    const nodes = this.__getNodes(components);
+
+    // TODO : render to screen between each hierarchical render in order to get the proper sizes.
+    // or compute the final sizes/length ahead of time
+    // (ie. without actually re-rendering w/ D3/nunjuncks)
+
+    // For each parent, from the deepest nodes up to the root, get a layout for its children
+    return Promise.all(
+      this.__getParentsByDepth(nodes)
+        .map(
+          (node) => this.____generateELKLayout(node, nodes, links),
+        ),
+    );
+  }
+
+  /**
+   * Write a generated layout to the corresponding graph.
+   * @param {ElkNode[]} layout - previously generated layout, see {@link __generateAllElkLayouts}.
+   * @private
+   */
+  __writeLayout(layout) {
+    layout.forEach((elkNode) => this.__writeSingleDepthLayout(elkNode));
+  }
+
+  /**
+   * Build a {@link NodeData} map from the components.
+   * @param {Component[]} components - input components.
+   * @returns {Map<string, NodeData>} a map of nodes.
+   * @private
+   */
+  __getNodes(components) {
+    // constructing our data objects from the given components
+    /** @type {Map<string,NodeData>} */
+    const nodes = new Map(components.map((component) => [component.id, {
+      raw: component, children: [], parent: null, depth: null,
+    }]));
+
+    // defines the root
+    /** @type {NodeData} */
+    const root = {
+      raw: null, children: [], parent: null, depth: 0,
```

Comment:
Instead of having an array for children, why not directly use a map ?

Reply:
I assume that you are talking about using a `Map` object for key-value relationship.

What's the point ? 

We are always iterating over all children and never looking for a specific child. (which is why we would use a `Map` object) 
And `Map` objects have no `.map( ...)` method to help for iteration.